### PR TITLE
Add type hints for base58.py

### PR DESCRIPTION
**Pull request by fanquake**
TLDR: This add type hints to base58.py, as well as some basic tests and a Travis job that runs [`mypy`](http://mypy-lang.org). 

This adds some simple base58 tests (copied from the Bitcoin Core repo), [type hints](https://docs.python.org/3.6/library/typing.html) for base58.py and a Travis job that runs `mypy` on base58.py. I started with base58.py to keep everything self contained.

`mypy` was added as a dev dependency using `poetry add mypy --dev`.

Given that the project is using > Python 3.6, we can also take advantage [variable type annotations](https://www.python.org/dev/peps/pep-0526/).

Just opening as a draft for some concept level discussion. If there is interested then more anntotations can be added, and the Travis `mypy` scope increased. If we are happy to add type annotations here, then I'll PR the same changes to any upstream libraries, i.e [`bech32.py`](https://github.com/sipa/bech32/blob/master/ref/python/segwit_addr.py).

Another good reference for type hints is: https://mypy.readthedocs.io/en/latest/cheat_sheet_py3.html

### CODE REVIEW

**Comment by jb55** [comment]
Approach ACK!

Can't go wrong with more type annotations.

**Comment by IlyasRidhuan** [comment]
Concept Ack

**Comment by Sjors** [comment]
@fanquake can you force push a trivial change when you're released from Travis prison? `git commit --amend --date "$(date)"` should do the trick.

**Comment by fanquake** [comment]
@Sjors I've force pushed a change.

**Comment by achow101** [comment]
The project is actually listed as allowing Python 3.5.6, so I believe this would break that. However I wouldn't be opposed to bumping the minimum to 3.6.

Also needs rebase.

**Comment by fanquake** [comment]
Will rebase once #246 and #247 are in.

**Comment by practicalswift** [comment]
Concept ACK: Python without type hints is like driving without a safety belt :-)

(Like C++ without the sanitisers or static analysis!)

**Comment by achow101** [comment]
Tried using Python 3.5.6, our current minimum, and got:

```
Traceback (most recent call last):
  File "./run_tests.py", line 7, in <module>
    from test_base58 import TestBase58
  File "/home/andy/bitcoin/HWI6/test/test_base58.py", line 12
    TEST_VECTORS: List[Tuple[str, str]] = [
                ^
SyntaxError: invalid syntax
```

File: .travis.yml [hunk 1]
```diff
@@ -54,6 +54,8 @@ jobs:
         -   name: lint
             stage: lint
             script: flake8
+        -   name: Type annotation checking
+            script: poetry run mypy hwilib/base58.py
```
Comment (achow101) [inline .travis.yml]: Could you add an `install` phase that just installs `mypy` and whatever else is needed to run the check?

Comment (fanquake) [reply .travis.yml]: Updated to use an `install` phase, and dropped the `mypy` dev dependency.

**Comment by fanquake** [comment]
> Tried using Python 3.5.6, our current minimum, and got:

If you're still happy to bump the minimum Python version to 3.6.x, I can do that here. Otherwise I can propose it in a separate PR?

**Comment by fanquake** [comment]
Rebased for #261 & #262.

**Comment by laanwj** [comment]
ACK b739f38435f3792ba00f45edf824db56e169ede1

**Comment by achow101** [comment]
ACK b739f38435f3792ba00f45edf824db56e169ede1
